src/dtype.rs

```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+    }
+
+    /// Returns true if the data type is a sub-array.
+    pub fn has_subarray(&self) -> bool {
+        // equivalent to PyDataType_HASSUBARRAY(self)
+        unsafe { !(*self.as_dtype_ptr()).subarray.is_null() }
+    }
+
+    /// Returns true if the data type is a structured type.
+    pub fn has_fields(&self) -> bool {
+        // equivalent to PyDataType_HASFIELDS(self)
+        unsafe { !(*self.as_dtype_ptr()).names.is_null() }
+    }
+
+    /// Returns true if the data type is unsized
+    pub fn is_unsized(&self) -> bool {
+        // equivalent to PyDataType_ISUNSIZED(self)
+        self.itemsize() == 0 && !self.has_fields()
+    }
+
+    /// Returns true if data type byteorder is native, or `None` if not applicable.
+    pub fn is_native_byteorder(&self) -> Option<bool> {
+        // based on PyArray_ISNBO(self->byteorder)
+        match self.byteorder() {
+            b'=' => Some(true),
+            b'|' => None,
+            byteorder if byteorder == NPY_BYTEORDER_CHAR::NPY_NATBYTE as u8 => Some(true),
+            _ => Some(false),
+        }
+    }
+
+    /// Returns an ordered list of field names, or `None` if there are no fields.
+    ///
+    /// The names are ordered according to increasing byte offset.
+    ///
+    /// Equivalent to [`np.dtype.names`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.names.html).
+    pub fn names(&self) -> Option<Vec<&str>> {
+        if !self.has_fields() {
+            return None;
+        }
+        let names = unsafe { PyTuple::from_borrowed_ptr(self.py(), (*self.as_dtype_ptr()).names) };
+        <_>::extract(names).ok()
+    }
+
+    /// Returns a dictionary of fields, or `None` if not a structured type.
+    ///
+    /// The dictionary is indexed by keys that are the names of the fields. Each entry in
+    /// the dictionary is a tuple fully describing the field: `(dtype, offset)`.
+    ///
+    /// Note: titles (the optional 3rd tuple element) are ignored.
+    ///
+    /// Equivalent to [`np.dtype.fields`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.fields.html).
+    pub fn fields(&self) -> Option<BTreeMap<&str, (&PyArrayDescr, usize)>> {
+        if !self.has_fields() {
+            return None;
+        }
+        // TODO: can this be done simpler, without the incref?
+        let dict = unsafe { PyDict::from_borrowed_ptr(self.py(), (*self.as_dtype_ptr()).fields) };
+        let mut fields = BTreeMap::new();
+        (|| -> PyResult<_> {
+            for (k, v) in dict.iter() {
+                // TODO: alternatively, could unwrap everything here
+                let name = <_>::extract(k)?;
+                let tuple = v.downcast::<PyTuple>()?;
+                let dtype = <_>::extract(tuple.as_ref().get_item(0)?)?;
+                let offset = <_>::extract(tuple.as_ref().get_item(1)?)?;
+                fields.insert(name, (dtype, offset));
+            }
+            Ok(fields)
+        })()
+        .ok()
```

Comment:
Why swallow up the `Result` into `Option` instead of returning `PyResult`?

Reply:
Because `Option` means there's no fields. So it would have to be `Option<PyResult<...>>`.

Thing is, this `PyResult<>` will be always ok, those fields are not mutable from outside, are created in the C code and are relied upon in the entire numpy.

Reply:
Then `unwrapping` seems appropriate. If not, `PyResult<Option<_>, _>` would be the more fitting signature IMHO. (Similary for the other methods which turn `PyResult` into `Option`.)

Reply:
I think unwrapping is ok here (and if anyone can make it panic from within Python, it should be a bug submitted to numpy upstream).

In which case, is an iterator or a collected (btreemap) preferable?

Reply:
Let's resolve this discussion in favour of https://github.com/PyO3/rust-numpy/pull/261/files#r788000640 to reduce the clutter.